pypdf/_writer.py

```diff
@@ -2246,15 +2246,48 @@ def remove_text(self, font_names: Optional[List[str]] = None) -> None:
             font_names = []
 
         for page in self.pages:
-            font_ids = []
-            fonts = page.get("/Resources", {}).get("/Font", {})
-            for font_id, font_info in fonts.items():
-                font_name = font_info.get("/BaseFont", "").split("+")[-1]
-                if font_name in font_names:
-                    font_ids.append(font_id)
+            resource_names_to_remove = []
+
+            # Content streams reference fonts and other resources with names like "/F1" or "/T1_0"
+            # Font names need to be converted to resource names for easier removal
+            if font_names:
+                # Recursively loop through the page to gather font info
+                def get_font_info(
+                    obj: Any,
+                    font_info: Optional[Dict[str, Any]] = None,
+                    key: Optional[str] = None
+                ) -> Dict[str, Any]:
+                    if font_info is None:
+                        font_info = {}
+                    if isinstance(obj, dict):
+                        if obj.get("/Type") == "/Font":
+                            font_name = obj.get("/BaseFont", "")
+                            # Normalize font names like "/RRXFFV+Palatino-Bold" to "Palatino-Bold"
+                            normalized_font_name = font_name.lstrip("/").split("+")[-1]
+                            if normalized_font_name not in font_info:
+                                font_info[normalized_font_name] = {
+                                    "normalized_font_name": normalized_font_name,
+                                    "resource_names": [],
+                                }
+                            if key not in font_info[normalized_font_name]["resource_names"]:
+                                font_info[normalized_font_name]["resource_names"].append(key)
+                        for k in obj:
+                            font_info = get_font_info(obj[k], font_info, k)
+                    elif isinstance(obj, (list, ArrayObject)):
+                        for child_obj in obj:
+                            font_info = get_font_info(child_obj, font_info)
+                    elif isinstance(obj, IndirectObject):
```

Comment:
Wouldn't it be more effective to do this check independently at the start of the nested function?

Reply:
I don't know a lot about PDF structure – is there a better way to check if it's an IndirectObject? I'm not sure if I understand your suggestion.

Reply:
There should not be any real need to know how indirect references work, but resolving it with `obj.get_object()` can be optimized.

We should be able to avoid one recursive call for each indirect reference by modifying the first lines of `get_font_info` like this:

```
                    if font_info is None:
                        font_info = {}
                    if isinstance(obj, IndirectObject):
                        obj = obj.get_object()
                    if isinstance(obj, dict):
```